File: app/lib/features/tracker/domain/user_procedure.dart
/// Status of a tracked procedure.
enum ProcedureStatus {
  notStarted,
  inProgress,
  completed;

  String get apiValue {
    switch (this) {
      case ProcedureStatus.notStarted:
        return 'not_started';
      case ProcedureStatus.inProgress:
        return 'in_progress';
      case ProcedureStatus.completed:
        return 'completed';
    }
  }

  static ProcedureStatus fromString(String value) {
    switch (value) {
      case 'not_started':
        return ProcedureStatus.notStarted;
      case 'in_progress':
        return ProcedureStatus.inProgress;
      case 'completed':
        return ProcedureStatus.completed;
      default:
        return ProcedureStatus.notStarted;
    }
  }
}

/// A user's tracked procedure.
class UserProcedure {
  const UserProcedure({
    required this.id,
    required this.procedureRefType,
    required this.procedureRefId,
    required this.title,
    required this.status,
    this.dueDate,
    this.notes,
    this.completedAt,
    this.createdAt,
  });

  final String id;
  final String procedureRefType;
  final String procedureRefId;
  final String title;
  final ProcedureStatus status;
  final DateTime? dueDate;
  final String? notes;
  final DateTime? completedAt;
  final DateTime? createdAt;

  factory UserProcedure.fromJson(Map<String, dynamic> json) {
    return UserProcedure(
      id: json['id'] as String,
      procedureRefType: json['procedure_ref_type'] as String? ?? '',
      procedureRefId: json['procedure_ref_id'] as String? ?? '',
      title: json['title'] as String? ?? '',
      status: ProcedureStatus.fromString(
          json['status'] as String? ?? 'not_started'),
      dueDate: json['due_date'] != null
          ? DateTime.tryParse(json['due_date'] as String)
          : null,
      notes: json['notes'] as String?,
      completedAt: json['completed_at'] != null
          ? DateTime.tryParse(json['completed_at'] as String)
          : null,
      createdAt: json['created_at'] != null
          ? DateTime.tryParse(json['created_at'] as String)
          : null,
    );
  }
}

File: backend/app_service/models/user_procedure.py
"""UserProcedure model — maps to the user_procedures table (DATA_MODEL.md §8)."""

import uuid
from datetime import date, datetime

from sqlalchemy import Date, DateTime, ForeignKey, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from database import Base


class UserProcedure(Base):
    """User's tracked procedure (Admin Tracker)."""

    __tablename__ = "user_procedures"

    id: Mapped[str] = mapped_column(
        String(36), primary_key=True, default=lambda: str(uuid.uuid4())
    )
    user_id: Mapped[str] = mapped_column(
        String(128), ForeignKey("profiles.id"), nullable=False
    )
    procedure_ref_type: Mapped[str] = mapped_column(
        String(10), nullable=False
    )  # 'admin' | 'visa'
    procedure_ref_id: Mapped[str] = mapped_column(String(36), nullable=False)
    status: Mapped[str] = mapped_column(
        String(20), nullable=False, default="not_started"
    )
    due_date: Mapped[date | None] = mapped_column(Date, nullable=True, default=None)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True, default=None)
    completed_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True, default=None
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now(),
        onupdate=func.now(),
    )
    deleted_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True, default=None
    )
